src/n-core/core/types/MonoBehaviourSingleton.cs

```diff
@@ -0,0 +1,73 @@
+using UnityEngine;
+
+namespace N {
+    /// <summary>
+    /// Extend this class if you want your monobehaviour to be a singleton!
+    /// Your child class should implement Init() rather than Awake()
+    /// </summary>
+    /// <typeparam name="T">The concrete type of your extending class</typeparam>
+    public abstract class MonoBehaviourSingleton<T> : MonoBehaviour where T : MonoBehaviourSingleton<T> {
+        static T instance;
+
+        /// <summary>
+        /// Whether or not this object should persist when loading new scenes.
+        /// This should be set in the child classes Init() method.
+        /// </summary>
+        bool persist = false;
+
+        /// <summary>
+        /// The instance of this Singleton Monobehaviour
+        /// </summary>
+        public static T Instance {
+            get {
+                // This would only EVER be null if some other MonoBehavior requests the instance
+                // in its' Awake method.
+                if(instance == null) {
+                    Debug.Log("[MonoBehaviourSingleton] Finding instance of '" + typeof(T).ToString() + 
+                            "' object.");
+                    instance = FindObjectOfType(typeof(T)) as T;
+                    // This should only occur if 'T' hasn't been attached to any game
+                    // objects in the scene.
+                    if(instance == null) {
+                        Debug.LogError("[MonoBehaviourSingleton] No instance of " + typeof(T).ToString()
```

Comment:
can we:

    if (typeof(IAutoSingleton).IsAssignableFrom(typeof(T))) {
      var gameObject = new GameObject();
      instance = gameObject.AddComponent<T>();
      instance.transform.name = instance.AutoSingletonName(); // <-- From IAutoSingleton
    }

That way we can have singleton objects that are automatically created if required with default values, rather than just bailing out with an error when the singleton hasn't been added to the scene?